src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java

```diff
@@ -42,17 +47,60 @@
 
     protected Name listArg;
 
-    private boolean cleaned = false;
-    private LdapResult res;
-    private LdapClient enumClnt;
     private Continuation cont;  // used to fill in exceptions
     private Vector<LdapEntry> entries = null;
     private int limit = 0;
     private int posn = 0;
-    protected LdapCtx homeCtx;
+
     private LdapReferralException refEx = null;
     private NamingException errEx = null;
 
+    /* This class maintains the pieces of state that need to be cleaned up (or
+     * are needed for cleanup). It gets registered with Cleaner to perform cleanup.
+     *
+     * reachabilityFences are used to ensure that an AbstractLdapNamingEnumeration
+     * instance does not become unreachable while one of its methods is still
+     * executing (possibly leading to EnumCtx being cleaned up while it's still in use).
+     *
+     * Because the state is mutable, fullFence() is used to ensure that changes
+     * made on the main/program thread are seen by the cleanup thread.
+     */
+    private static class EnumCtx implements Runnable {
+        private LdapCtx homeCtx;
+        private LdapResult res;
+        private LdapClient enumClnt;
+
+        private EnumCtx(LdapCtx homeCtx, LdapResult answer, LdapClient client) {
+            this.homeCtx = homeCtx;
+            this.res = answer;
+            this.enumClnt = client;
+        }
+
+        @Override
+        public void run() {
+            // Ensure changes on the main/program thread happens-before cleanup
+            VarHandle.fullFence();
+
+            if (enumClnt != null) {
+                if (homeCtx != null) {
+                    enumClnt.clearSearchReply(res, homeCtx.reqCtls);
+                }
+                enumClnt = null;
+            }
+            if (homeCtx != null) {
+                homeCtx.decEnumCount();
+                homeCtx = null;
+            }
```

Comment:
Cleaner's task (run() method) is guaranteed to be called at most once. So there's no need for above "idempotent" logic. Unless some of those fields are optional when the instance is constructed....